---
title: "WGCNA_consensus"
author: "Karissa Barthelson"
date: "2024-06-17"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  autodep = TRUE,
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center", 
  out.width ="75%", 
  out.height = "75%"
)
```

```{r}
library(tidyverse)
library(magrittr)
library(readxl)

# vis
library(pander)
library(kableExtra)
library(UpSetR)
library(scales)
library(pheatmap)
library(ggpubr)
library(ggrepel)
library(ggfortify)
library(ggforce)
library(RColorBrewer)
library(colorspace)
library(igraph)

# analysis
library(WGCNA)
library(topGO)
library(org.Dr.eg.db)
library(clusterProfiler)
library(KEGGREST)
library(msigdbr)
library(matrixStats)

theme_set(theme_classic())
options(stringsAsFactors = FALSE)
allowWGCNAThreads()  
```
# Introduction 

In this page, I will create a consensus network of the male and female brains. This will allow me to compare modules that are present in the brains of both sexes to find the ones which are present in both. 

# Read in data
First, read in the logCPM expression data (after CQN normalisation) processed in the [RNAseq](RNA_seq.html) analysis. 
```{r}
# read in data 
x <- readRDS("data/RNAseq/R-objects/x_dgeObject_withCQNoffset.rds")

# update the group factor for this analysis
x$samples %<>% 
  mutate(group = paste0(genotype, "_", treatment2) %>% 
           as.factor())

# CQN adjusted logCPM
logCPM <- readRDS("data/RNAseq/R-objects/logCPM_CQN")

# make vectors for subsetting
male.fish.ids <- x$samples %>% 
  dplyr::filter(sex == "male") %>% 
  .$fish_id

female.fish.ids <- x$samples %>% 
  dplyr::filter(sex == "female") %>% 
  .$fish_id

# subset DGEs
x.males <- x[,male.fish.ids]
x.females <- x[,female.fish.ids]

# subset logCPMs
logCPM.male <- logCPM[,male.fish.ids]
logCPM.female <- logCPM[,female.fish.ids]

# set label names 
setLabels = c("female brains", "male brains")
```

## prepare for WGCNA

I will filter the genes to retain only the most variable genes. This is because genes which do not vary across samples will not be informative for network analysis, and will just increase the memory needed to process, and wont provide any biologically meaningful information when defining modules. 

The code used to produce the consensus network is largely based on the WGCNA tutorial (ask me for the pdfs, since the original website has disappeared). 

```{r}
# we are working with 2 expression sets
nSets = 2

# Retain genes with sufficiently high variation in expression across all samples.
# Here we define sufficiently high variation as having a standard deviation above the 25th quantile (so not in bottom 25%).

variable.genes <- rowSds(logCPM) >  quantile(rowSds(logCPM), 0.25)


# peform the filtering
logCPM.female.filtered <- logCPM.female %>% extract(variable.genes,)
logCPM.male.filtered <- logCPM.male %>% extract(variable.genes,)


# Form multi-set expression data: 
multiExpr = vector(mode = "list", length = nSets)

# add male and female datasets
multiExpr[[1]] = list(data = as.data.frame(t(logCPM.female.filtered)))
multiExpr[[2]] = list(data = as.data.frame(t(logCPM.male.filtered)))

# Check that the data has the correct format for many functions operating on multiple sets:
exprSize = checkSets(multiExpr)

# Check that all genes and samples have sufficiently low numbers of missing values.
gsg = goodSamplesGenesMS(multiExpr, verbose = 3)
gsg$allOK

# not all genes are OK, so remove these using the following function. 
if (!gsg$allOK)
{
  # Print information about the removed genes:
  if (sum(!gsg$goodGenes) > 0)
    printFlush(paste("Removing genes:", paste(names(multiExpr[[1]]$data)[!gsg$goodGenes],
                                              collapse = ", ")))
  for (set in 1:exprSize$nSets)
  {
    
    if (sum(!gsg$goodSamples[[set]]))
      printFlush(paste("In set", setLabels[set], "removing samples",
                       paste(rownames(multiExpr[[set]]$data)[!gsg$goodSamples[[set]]], collapse = ", ")))
    # Remove the offending genes and samples
    multiExpr[[set]]$data = multiExpr[[set]]$data[gsg$goodSamples[[set]], gsg$goodGenes];
  }
  # Update exprSize
  exprSize = checkSets(multiExpr)
}
```

## sample clustering

We now cluster the samples on their complete distance, separately in each set.

```{r}
sampleTrees = list()
for (set in 1:nSets)
{
sampleTrees[[set]] = hclust(dist(multiExpr[[set]]$data), method = "average")
}

par(mfrow=c(2,1))
par(mar = c(0, 4, 2, 0))
for (set in 1:nSets)
plot(sampleTrees[[set]], main = paste("Sample clustering on all genes in", setLabels[set]),
xlab="", sub="", cex = 0.7);
```


## load fish metadata
```{r}
# Form a multi-set structure that will hold the fish metadata 
Traits = vector(mode = "list", length = nSets)

Traits[[1]]$data = x.females$samples
Traits[[2]]$data = x.males$samples

# Define data set dimensions
nGenes = exprSize$nGenes
nSamples = exprSize$nSamples
```

# Step-by-step network construction and module detection

## Choice of soft therholding power (beta)

The plot below shows the summary network indices (y-axes) as functions of the soft thresholding power (x-axes). Numbers in the plots below indicate the corresponding soft thresholding powers. The plots indicate that approximate scale-free topology is attained around the soft-thresholding power of 14 for both sets. Because the summary connectivity measures decline steeply with increasing soft-thresholding power, it is advantageous to choose the lowest power that satisfies the approximate scale-free topology criterion.

```{r}
# Choose a set of soft-thresholding powers
powers = c(seq(4,10,by=1), seq(12,20, by=2))

# Initialize a list to hold the results of scale-free analysis
powerTables = vector(mode = "list", length = nSets)

# Call the network topology analysis function for each set in turn
for (set in 1:nSets)

  powerTables[[set]] = list(
    data = pickSoftThreshold(multiExpr[[set]]$data,
                             networkType = "signed",
                             powerVector = powers,
                             verbose = 2)[[2]]
  )


# Add the sex to help with plotting. 

powerTables[[1]]$data %<>% 
  mutate(sex = "female")

powerTables[[2]]$data %<>% 
  mutate(sex = "male")

# plot the resutls 
ggarrange( 
  # ~~~ Scale Free Topology Model Fit ~~
  powerTables[[1]]$data %>% 
  bind_rows(powerTables[[2]]$data) %>% 
  ggplot(
    aes(x = Power, y = -sign(slope) * SFT.R.sq, 
        colour = sex)
    ) +
    geom_label(
      aes(label = Power), 
      show.legend = F,
      fill = NA
    ) +
    geom_hline(
      yintercept = 0.8, 
      colour = "darkgrey") +
    annotate(
      geom = "text", 
      label = "beta = 0.8",
      colour = "darkgrey",
      x = 4, 
      y = 0.78
    ) +
    labs(
      x = "Soft Threshold (power)",
      y = "Scale Free Topology Model Fit,\nsigned R^2",
      title = "Scale independence"
    ), 
  
  # ~~~ Mean connectivity  ~~~
  
  powerTables[[1]]$data %>% 
  bind_rows(powerTables[[2]]$data) %>% 
    ggplot(
      aes(x = Power, y = mean.k., colour = sex)
    ) +
    geom_label(
      aes(label = Power), 
      fill = NA, 
      #show.legend = F
    ) +
    geom_hline(
      yintercept = 0, 
      colour = "darkgrey") +
    annotate(
      geom = "text", 
      label = "k = 0",
      colour = "darkgrey",
      x = 3, 
      y = 100
    ) +
    labs(
      x = "Soft Threshold (power)",
      y = "Mean connectivity (k)",
      title = "Mean connectivity" 
  ), 
  
  # ~~~ Median connectivity  ~~~
  
  powerTables[[1]]$data %>% 
    bind_rows(powerTables[[2]]$data) %>% 
    ggplot(
      aes(x = Power, y =  median.k., colour = sex)
    ) +
    geom_label(
      aes(label = Power), 
      fill = NA, 
      show.legend = F
    ) +
    geom_hline(
      yintercept = 0, 
      colour = "darkgrey") +
    annotate(
      geom = "text", 
      label = "k = 0",
      colour = "darkgrey",
      x = 3, 
      y = 100
    ) +
    labs(
      x = "Soft Threshold (power)",
      y = "Median connectivity (k)",
      title = "Median connectivity" 
  ), 
  
    # ~~~ Max connectivity  ~~~
  
  powerTables[[1]]$data %>% 
    bind_rows(powerTables[[2]]$data) %>% 
    ggplot(
      aes(x = Power, y =  max.k., colour = sex)
    ) +
    geom_label(
      aes(label = Power), 
      fill = NA, 
      show.legend = F
    ) +
    geom_hline(
      yintercept = 0, 
      colour = "darkgrey") +
    annotate(
      geom = "text", 
      label = "k = 0",
      colour = "darkgrey",
      x = 3, 
      y = 100
    ) +
    labs(
      x = "Soft Threshold (power)",
      y = "Max connectivity (k)",
      title = "Max connectivity" 
  ), 
  
  common.legend = TRUE
)
```

## Adjacnecy matrices
Network construction starts by calculating the adjacencies in the individual sets, using the soft thresholding power. 

```{r}
softPower = 14

# Initialize an appropriate array to hold the adjacencies
adjacencies = array(0, dim = c(nSets, nGenes, nGenes))

# set this higher so I dont run out of memory
Sys.setenv('R_MAX_VSIZE'=32000000000)

# Calculate adjacency in each individual data set
for (set in 1:nSets)
adjacencies[set, , ] = abs(cor(multiExpr[[set]]$data, use = "p"))^softPower
```

## Calculation of Topological Overlap Matrices (TOM) 
We now turn the adjacencies into Topological Overlap Matricies (TOM). 

```{r}
# Initialize an appropriate array to hold the TOMs
TOM = array(0, dim = c(nSets, nGenes, nGenes))

# Calculate TOMs in each individual data set
for (set in 1:nSets)
TOM[set, , ] = TOMsimilarity(adjacencies[set, , ])
```

## Scaling of Topological Overlap Matrices to make them comparable across sets

Topological Overlap Matrices of different data sets may have different statistical properties. For example, the TOM in the male data may be systematically lower than the TOM in female data. Since consensus is defined as the component-wise minimum of the two TOMs, a bias may result. Here we illustrate a simple scaling that mitigates the effect of different statistical properties to some degree. We scale the male TOM such that the 95th percentile equals the 95th percentile of the female TOM. 

```{r}
# Define the reference percentile
scaleP = 0.95

# Set RNG seed for reproducibility of sampling
set.seed(12345)

# Sample sufficiently large number of TOM entries
nSamples = as.integer(1/(1-scaleP) * 1000);

# Choose the sampled TOM entries
scaleSample = sample(nGenes*(nGenes-1)/2, size = nSamples)

TOMScalingSamples = list();

# These are TOM values at reference percentile
scaleQuant = rep(1, nSets)

# Scaling powers to equalize reference TOM values
scalePowers = rep(1, nSets)

# Loop over sets
for (set in 1:nSets)
{
  # Select the sampled TOM entries
  TOMScalingSamples[[set]] = as.dist(TOM[set, , ])[scaleSample]
  # Calculate the 95th percentile
  scaleQuant[set] = quantile(TOMScalingSamples[[set]],
                             probs = scaleP, type = 8);
  # Scale the male TOM
  if (set>1)
  {
    scalePowers[set] = log(scaleQuant[1])/log(scaleQuant[set]);
    TOM[set, ,] = TOM[set, ,]^scalePowers[set];
  }
}


# For plotting, also scale the sampled TOM entries
scaledTOMSamples = list()

for (set in 1:nSets)
  scaledTOMSamples[[set]] = TOMScalingSamples[[set]]^scalePowers[set]

# Open a suitably sized graphics window
sizeGrWindow(6,6)

# qq plot of the unscaled samples
qqUnscaled = qqplot(TOMScalingSamples[[1]], TOMScalingSamples[[2]], 
                    plot.it = TRUE, cex = 0.6,
                    xlab = paste("TOM in", setLabels[1]), 
                    ylab = paste("TOM in", setLabels[2]),
                    main = "Q-Q plot of TOM", pch = 20)

# qq plot of the scaled samples
qqScaled = qqplot(scaledTOMSamples[[1]], scaledTOMSamples[[2]], plot.it = FALSE)
points(qqScaled$x, qqScaled$y, col = "red", cex = 0.6, pch = 20);
abline(a=0, b=1, col = "blue")

legend("topleft", legend = c("Unscaled TOM", "Scaled TOM"), pch = 20, col = c("black", "red"))
```

Note that my qq plot looks significantly different from the TOM in the tutorial. I really cant find much info online about how important this is. 
  
## Calculation of consensus Topological Overlap matrix

We now calculate the consensus Topological Overlap by taking the component-wise (“parallel”) minimum of the TOMs in individual sets. Thus, the consensus topological overlap of two genes is only large if the corresponding entries in the two sets are also large. 

```{r}
consensusTOM = pmin(TOM[1, , ], TOM[2, , ])
```

# Clustering and module identification
We use the consensus TOM as input to hierarchical clustering, and identify modules in the resulting dendrogram using the Dynamic Tree Cut algorithm . 

```{r}
# Clustering
consTree = hclust(as.dist(1-consensusTOM), method = "average")

# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 30

# Module identification using dynamic tree cut:
unmergedLabels = 
  cutreeDynamic(
    dendro = consTree, 
    distM = 1-consensusTOM,
    
    method = "hybrid",
    deepSplit = 2, cutHeight = 0.995,
    minClusterSize = minModuleSize,
    pamRespectsDendro = FALSE 
  )

unmergedColors = labels2colors(unmergedLabels)

table(unmergedLabels)

plotDendroAndColors(
  consTree, unmergedColors, "Dynamic Tree Cut",
  dendroLabels = FALSE, hang = 0.03,
  addGuide = TRUE, guideHang = 0.05
  )

```

## merge similar modules
```{r}
merge = mergeCloseModules(multiExpr, unmergedLabels, cutHeight = 0.25, verbose = 3)

# Numeric module labels
moduleLabels = merge$colors

# Convert labels to colors
moduleColors = labels2colors(moduleLabels)

# Eigengenes of the new merged modules:
consMEs = merge$newMEs;
```

There are only `r unique(moduleColors) %>% length()` consensus modules after merging. 

```{r}
table(moduleColors)
```

These are then plotted under the dendrogram. This means that in the consensus network, not many similar genes/modules are present. 

```{r}
plotDendroAndColors(consTree, cbind(unmergedColors, moduleColors),
                    c("Unmerged", "Merged"),
                    dendroLabels = FALSE, 
                    #hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05
                    )
```

# enrichment analysis of the modules

I next perform GO analysis on the consensus modules to determine whether they are biologically meaningful. I will do this by hypergeometric tests using the `enrichGO` function of [clusterProfiler](https://yulab-smu.top/biomedical-knowledge-mining-book/index.html) 

```{r}
# Make a list with the gene expression by module 
colour.names <- moduleColors %>% 
  unique

# make a list containing the modules and their respective genes
genesInModuleList <- 
 colour.names %>% 
  sapply(function(x) {
    logCPM.male.filtered[moduleColors==x,] %>% 
      rownames
  }, simplify = F)

# perform the enrichment analysis on GO terms. 
consensus.modules.enrichment.results <- 
  genesInModuleList %>% 
  sapply(function(y) {
      enrichGO(
        gene = y,
        universe = rownames(logCPM.male.filtered),
        OrgDb = org.Dr.eg.db,
        keyType = "ENSEMBL",
        ont = "ALL",
        pAdjustMethod = "BH",
        pvalueCutoff  = 0.01,
        qvalueCutoff  = 0.05,
        readable = TRUE)
  }, simplify = F)

# make a nice table for the results. 
consensus.modules.enrichment.results.tables <- 
  consensus.modules.enrichment.results %>% 
      lapply(function(x) {
        x %>% 
          as_tibble 
      }) %>% 
      imap(~ .x %>% mutate(module = .y)) # add mod name from list element name
```

# export consensus data

This will be used later to compare with the male and female networks. 

```{r}
save(consMEs, moduleColors, moduleLabels, consTree, Traits, multiExpr, consensus.modules.enrichment.results.tables, 
     file = "output/WGCNA/R_objects/Consensus-NetworkConstruction-man.RData") 
```
